src/main/java/techreborn/blockentity/machine/tier0/block/blockbreaker/BlockBreakerProcessor.java

```diff
@@ -0,0 +1,207 @@
+import techreborn.blockentity.machine.tier0.block.MachineProcessException;
+import techreborn.blockentity.machine.tier0.block.ProcessingStatus;
+
+import java.util.UUID;
+
+/**
+ * <b>Class handling the process of breaking a block</b>
+ * <br>
+ * The main purpose of this class is to implement the {@link #onTick(World, BlockPos)}.
+ * This method defines the logic for breaking a block
+ *
+ * @author SimonFlapse
+ * @see techreborn.blockentity.machine.tier0.block.BlockBreakerBlockEntity
+ */
+public class BlockBreakerProcessor extends BlockBreakerNbt {
+
+	private UUID processorId = UUID.randomUUID();
+	private BlockProcessable processable;
+
+	private int outputSlot;
+	private int fakeInputSlot;
+
+	private int baseBreakTime;
+	private int baseCostToBreak;
+
+	public BlockBreakerProcessor(BlockProcessable processable, int outputSlot, int fakeInputSlot, int baseBreakTime, int baseCostToBreak) {
+		this.processable = processable;
+
+		this.outputSlot = outputSlot;
+		this.fakeInputSlot = fakeInputSlot;
+
+		this.baseBreakTime = baseBreakTime;
+		this.baseCostToBreak = baseCostToBreak;
+	}
+
+	public ProcessingStatus getStatusEnum() {
+		return status;
+	}
+
+	public int getProgress() {
+		return (int) (((double) getCurrentBreakTime() / (double) getBreakTime()) * 100);
+	}
+
+	public ProcessingStatus onTick(World world, BlockPos positionInFront) {
+		try {
+			handleBlockBreakingProgressReset(world, positionInFront);
+
+			ensureRedstoneEnabled();
+
+			handleInterrupted();
+
+			ItemStack outputItemStack = processable.getInventory().getStack(outputSlot);
+
+			BlockState blockInFront = world.getBlockState(positionInFront);
+
+			handleBlockInFrontRemoved(blockInFront);
+
+			Item currentBreakingItem = processable.getInventory().getStack(fakeInputSlot).getItem();
+
+			ItemStack item = blockInFront.getBlock().asItem().getDefaultStack().copy();
+
+			ItemStack fakeItem = item.copy();
+
+			if (fakeItem.isOf(Items.AIR)) {
+				currentBreakingItem = null;
+			}
+
+			processable.getInventory().setStack(fakeInputSlot, fakeItem);
+
+			float hardness = BlockProcessorUtils.getHardness(world, blockInFront, positionInFront);
+
+			ensureBlockCanBeBroken(blockInFront, fakeItem, hardness);
+
+			this.breakTime = BlockProcessorUtils.getProcessTimeWithHardness(processable, baseBreakTime, hardness);
+
+			ensureBlockNotReplaced(currentBreakingItem, item);
+
+			ensureBlockFitInOutput(outputItemStack, item);
+
+			increaseBreakTime(world, positionInFront);
+
+			BlockProcessorUtils.playSound(processable, currentBreakTime);
+
+			breakBlock(world, positionInFront, outputItemStack, item);
+
+			status = BlockBreakerStatus.PROCESSING;
+		} catch (MachineProcessException ex) {
+			status = ex.getStatus();
+		}
+
+		return status;
+	}
+
+	private void ensureRedstoneEnabled() {
+		if (!processable.isActive(RedstoneConfiguration.RECIPE_PROCESSING)) {
+			throw new MachineProcessException(BlockBreakerStatus.IDLE_PAUSED);
```

Comment:
Using exceptions for control logic can be a cause for peformance issues, as when the exception is created it generates the stack trace.

See:

https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Throwable.java#L271
and
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Throwable.java#L794


A way to solve this is to override `fillInStackTrace` in MachineProcessException

```java
@Override
public synchronized Throwable fillInStackTrace() {
    return this;
}
```

Using Exception's for control flow isnt generally a good practice in software dev, I dont think its worth changing here though, when you have the above change.

Reply:
Good point, I read up on it and I agree it is an anti-pattern. (You learn something new everyday) 

I will try to change the control flow to not use exceptions, otherwise I'll make use of your suggestion.